ERCS/erc-7715.md

```diff
@@ -0,0 +1,480 @@
+
+#### Signers
+
+```tsx
+// A wallet is the signer for these permissions
+// `data` is not necessary for this signer type as the wallet is both the signer and grantor of these permissions
+type WalletSigner = {
+  type: "wallet";
+  data: {};
+};
+
+// A signer representing a single key.
+// `id` is a did:key identifier and can therefore represent both Secp256k1 or Secp256r1 keys, among other key types.
+type KeySigner = {
+  type: "key";
+  data: {
+    id: string;
+  };
+};
+
+// A signer representing a multisig signer.
+// Each element of `ids` is a did:key identifier just like the `key` signer.
+type MultiKeySigner = {
+  type: "keys";
+  data: {
+    ids: string[];
+  };
+};
+
+// An account that can be granted with permissions as in ERC-7710.
+type AccountSigner = {
+  type: "account";
+  data: {
+    id: `0x${string}`;
+  };
+};
+```
+
+### Permissions Types
+
+```tsx
+// Native token spending limit, e.g. ETH on Ethereum
+type NativeTokenLimitPermission = {
+  type: "native-token-limit";
+  data: {
+    amount: "0x..."; // wei in hex
+  };
+};
+
+// ERC20 spending limit
+type ERC20LimitPermission = {
+  type: "erc20-limit";
+  data: {
+    erc20Address: "0x..."; // ERC20 token address
+    amount: "0x..."; // wei in hex
+  };
+};
+
+// Gas spending limit
+type GasLimitPermission = {
+  type: "gas-limit";
+  data: {
+    amount: "0x..."; // wei in hex
+  };
+};
+
+// Allowlist of contract calls
+type ContractCallPermission = {
+  type: "contract-call";
+  data: {
+    // TODO
+  };
+};
+
+// The number of calls the session can make during each interval
+type RateLimitPermission = {
+  type: "rate-limit";
+  data: {
+    count: number; // the number of times during each interval
+    interval: number; // in seconds
+  };
+};
+```
+
+### Wallet-managed Sessions
+
+If the signer is specified as `wallet`, then the wallet itself manages the session. If the wallet approves the request, it MUST accept [ERC-5792](./eip-5792.md)’s `wallet_sendCalls` with the `permissions` capability, which MAY include the session with a `permissionsContext`. For example:
+
+```tsx
+[
+  {
+    version: "1.0",
+    chainId: "0x01",
+    from: "0xd46e8dd67c5d32be8058bb8eb970870f07244567",
+    calls: [
+      {
+        to: "0xd46e8dd67c5d32be8058bb8eb970870f07244567",
+        value: "0x9184e72a",
+        data: "0xd46e8dd67c5d32be8d46e8dd67c5d32be8058bb8eb970870f072445675058bb8eb970870f072445675",
+      },
+      {
+        to: "0xd46e8dd67c5d32be8058bb8eb970870f07244567",
+        value: "0x182183",
+        data: "0xfbadbaf01",
+      },
+    ],
+    capabilities: {
+      permissions: {
+        context: "<permissionsContext>",
+      },
+    },
```

Comment:
Can we also add a field for `capabilities.permissions.signature`? This way the app can sign things with their session key and easily pass this to the wallet for validation and userOp bundling along with the `context`. Passing an identifier for the session is definitely needed, but the wallet receiving `sendCalls` will also need a signature that provides per-userOp validation.

```suggestion
    capabilities: {
      permissions: {
        context: "<permissionsContext>",
        signature: "<optional signature from session key>"
      },
    },
```

Reply:
Conversation thread from telegram [here](https://t.me/c/2006770962/1136).

There are currently two paths in 7715:
1. App does everything without the wallet (7679/7710)
2. App leverages wallet connection in order to send transactions

This comment's proposal is a nuance of 2:
2a) wallet provisions session keys for app
2b) app brings its own session key

For modularity of product offerings, I think 2b is very important to support.

Consider a 2x2 grid with one axis for signing done by wallet/app and other axis for userOp submission by wallet/app.

|  | wallet signs | app signs |
| -- | -- | -- |
| wallet submits |  supported | _missing_ |
| app submits     | signTransaction | 7679 |

Wallets are able to prompt signing and submit userOps by definition.

If apps want maximum control they can own signing and use 7679 to submit userOps.

Something is missing to let apps own signing, but lean on the wallet for userOp submission.

To enable this missing box of app signs + wallet submits, we need to pass a signature from app to wallet and therefore I think this new and optional `capabilities.permissions.signature` field is the simplest approach.